### ISSUE TITLE
Act(Async) on valid null models is not possible

### ISSUE THREAD

**Issue by Corniel** [Qowaiv/qowaiv-validation#33]
Given the following code:

``` C#
return Initial()
    .Act(m => Result.For<MyModel>(null))
    .Act(m => m ?? new MyModel())
    .Act(m => throw new NotSupportedException());
```
The `NotSupportedException` is never thrown, nor is the `m ?? new MyModel()` executed.

Reason is, that `Act` is not executed if the `Result<T>` is invalid, or if the `Value` is `null`. I think that was a bad design choice, that should be changed. I see two options:
1. If the value is null and the result is valid, add an error to `Result<T>` (so it becomes invalid)
2. Just allow `Act` to be executed on `null` values.

**Comment by pmdevers** [comment]
I would go with returning an invalid result because allowing null could cause some unexpected side effects down the chain.

**Comment by LauraXiulan** [comment]
Option 2 means that you would always need a null check in the next act, but it does allow you to continue with a null return. Option 1 does not, but it does make it clearer that something in the chain is failing as opposed to what it does currently, which is just nothing at all.

**Comment by LauraXiulan** [comment]
I also think option 1 is a safer choice, even though I would sometimes like to continue with a null result.

**Comment by pmdevers** [comment]
Implement ActOnNull or something.

**Comment by ROSDEV** [comment]
Always feel like null being a odd one. 
Is a ActOnNull feasible or overloading with a 'OnNull' action ?

**Comment by jkoene** [comment]
null is often used to return when something isn't found. It has its function, but that doesn't mean the intention is always clear. And it could be (accidentally) implemented in a wrong way. 

It depends on what qowaiv wants to enforce on what to use. But it goes a bit deeper.

On option 1:
I don't think you should suddenly make a 'valid' result 'invalid' and add a error making it impossible for the code to continue on the value which was deemed valid. 
It would make it more clear/safer but then making a Result.For<MyModel>(null)) should not be possible at all and you would need a NullObject Pattern (https://en.wikipedia.org/wiki/Null_object_pattern) 

On option 2:
Although it feels more 'dangerous' to just continue, this is where the developer is responsible, the developer should know if it is possible that null is returned and then act on it. With the implementation of option 2 you at least make it possible for the developer to act on it. And for example create a new value or create a invalid response with error message. 

Implementing a ActOnNull (or all) for me doesn't make much sense, because the developer is still responsible for handling it properly it won't add additional value / make it less clear when to use what function and maybe you end up with people only calling ActOnNull which would make it ugly (in my opinion)

Overall my feeling goes for option 2.

**Comment by Corniel** [comment]
@jkoene When option 1 would be choosen: Result.For(null) would lead to a result not being valid.

**Comment by mars888** [comment]
I think option 1 would be the most reasonable choice, assuming the validity of the following argument (the code is quite new to me, so I'm possibly making some wrong assumptions on the implementation here):

Currently a `Result<T>` represents a validation for a model of type `T`; or for all possible instances of `T`. The type `T` itself can never change for a given result (there is no `Select` functionality as in Linq to go from `Result<T>` to `Result<U>`). Allowing null would therefor mean that null becomes part of the set of all possible instances of `T`. But in, my experience, this is generally not what you are trying to model. Mostly `T` being allowed to be null would mean that there is a possibility for the absence of a `T`.

Modelling the potential absence of a `T` instance can be achieved with option 1 by using a null-object or something like a `Maybe<T>`/`Optional<T>` implementation. So with option 1 all cases can still be covered, even making the special case of absence more explicit, while at the same time safe-guarding against accidental nulls at the same time.

**Comment by Corniel** [comment]
> Currently a `Result<T>` represents a validation for a model of type `T`; or for all possible instances of `T`. The type `T` itself can never change for a given result (there is no `Select` functionality as in Linq to go from `Result<T>` to `Result<U>`). 

That is not true. You can do something like this:

``` C#
Result<int> result = Result.For("12").Act(s => int.Parse(s));
```
> Modelling the potential absence of a `T` instance can be achieved with option 1 by using a null-object or something like a `Maybe<T>`/`Optional<T>` implementation.

The downside of an `Optional<T>` is that is not as elegant as using the `.?` and `??` syntax sugar to deal with option results. Furthermore, .NET has none available (that is not fully true, it is used in Roslyn.NET, but is not a dependency I want to add).

Defining it in Qowaiv.Validation.Abstractions is not the logical place, and adding any dependencies to an **Abstractions** package is not per se something to look forward too.

That being said, I see a lot of valuable input from all of you. The only thing that is clear, is that we agree on that the current implementation is wrong. That's a start.

**Comment by Corniel** [comment]
I also started a pull request (for now just to show/detect the places touched, and that tests fail due to this changed behaviour).

**Comment by Corniel** [comment]
What about this option?

``` C#
var nil = Result.Null<MyClass>(); // creates a valid result with null, as it explicit
var other = Result.For<MyClass>(null); // creates an invalid result.
```

So, by default, `null`-values become invalid, but it is still possible to create a valid null value if you really (really?, really!) want it.

**Comment by mars888** [comment]
> ```cs
> Result<int> result = Result.For("12").Act(s => int.Parse(s));
> ```

I indeed missed something here, it seems I got confused due to this example:
```csharp
            var result =
                Result.For("12")
                      .Act(s => int.TryParse(s, out var i)
                               ? Result.For(i)
                               : Result.WithMessages(
                                   ValidationMessage.Error("No parse")));
```
`result` here has type `Result<string>` as the `TryParse` expression yields a `Result` instead of a `Result<int>`. This in turn causes an overload of `Act` to be called that ignores the result value returned from the lambda. Is this the expected behavior here?

> The downside of an `Optional<T>` is that is not as elegant as using the `.?` and `??` syntax sugar to deal with option results. Furthermore, .NET has none available (that is not fully true, it is used in Roslyn.NET, but is not a dependency I want to add).
> 
> Defining it in Qowaiv.Validation.Abstractions is not the logical place, and adding any dependencies to an **Abstractions** package is not per se something to look forward too.

Completely agree with both points. But personally I still feel that allowing nulls may be a slippery slope (based on personal experience with result and option types).

**Comment by Corniel** [comment]
If you use `Result.WithMessages<int>`, that will not happen.

**Comment by mars888** [comment]
Indeed, but it's more about this being a bit of a trap that one can fall into without noticing. Apparently, I did at least 😂

**Comment by Corniel** [comment]
That is because `Result<T>` inherits form `Result`, I don't think you want (or can now) break that contract, nor would it be desirable that you change the API for Result.WithMessages() or Result.WithMessages<T>(), so that they can not be confused. Although, in it self, that is a totally different discussion anyway.

**Comment by Corniel** [comment]
Solved with #33.
